internal/graph/lookupresources2.go

```diff
@@ -0,0 +1,783 @@
+}
+
+func (crr *CursoredLookupResources2) LookupResources2(
+	req ValidatedLookupResources2Request,
+	stream dispatch.LookupResources2Stream,
+) error {
+	if req.TerminalSubject == nil {
+		return spiceerrors.MustBugf("no terminal subject given to lookup resources dispatch")
+	}
+
+	if slices.Contains(req.SubjectIds, tuple.PublicWildcard) {
+		return NewWildcardNotAllowedErr("cannot perform lookup resources on wildcard", "subject_id")
+	}
+
+	if len(req.SubjectIds) == 0 {
+		return spiceerrors.MustBugf("no subjects ids given to lookup resources dispatch")
+	}
+
+	// Sort for stability.
+	sort.Strings(req.SubjectIds)
+
+	ctx := stream.Context()
+	limits := newLimitTracker(req.OptionalLimit)
+	ci, err := newCursorInformation(req.OptionalCursor, limits, dispatchVersion)
+	if err != nil {
+		return err
+	}
+
+	return withSubsetInCursor(ci,
+		func(currentOffset int, nextCursorWith afterResponseCursor) error {
+			// If the resource type matches the subject type, yield directly as a one-to-one result
+			// for each subjectID.
+			if req.SubjectRelation.Namespace == req.ResourceRelation.Namespace &&
+				req.SubjectRelation.Relation == req.ResourceRelation.Relation {
+				for index, subjectID := range req.SubjectIds {
+					if index < currentOffset {
+						continue
+					}
+
+					if !ci.limits.prepareForPublishing() {
+						return nil
+					}
+
+					err := stream.Publish(&v1.DispatchLookupResources2Response{
+						Resource: &v1.PossibleResource{
+							ResourceId:    subjectID,
+							ForSubjectIds: []string{subjectID},
+						},
+						Metadata:            emptyMetadata,
+						AfterResponseCursor: nextCursorWith(index + 1),
+					})
+					if err != nil {
+						return err
+					}
+				}
+			}
+			return nil
+		}, func(ci cursorInformation) error {
+			// Once done checking for the matching subject type, yield by dispatching over entrypoints.
+			return crr.afterSameType(ctx, ci, req, stream)
+		})
+}
+
+func (crr *CursoredLookupResources2) afterSameType(
+	ctx context.Context,
+	ci cursorInformation,
+	req ValidatedLookupResources2Request,
+	parentStream dispatch.LookupResources2Stream,
+) error {
+	dispatched := &syncONRSet{}
+
+	// Load the type system and reachability graph to find the entrypoints for the reachability.
+	ds := datastoremw.MustFromContext(ctx)
+	reader := ds.SnapshotReader(req.Revision)
+	_, typeSystem, err := typesystem.ReadNamespaceAndTypes(ctx, req.ResourceRelation.Namespace, reader)
+	if err != nil {
+		return err
+	}
+
+	rg := typesystem.ReachabilityGraphFor(typeSystem)
+	entrypoints, err := rg.OptimizedEntrypointsForSubjectToResource(ctx, &core.RelationReference{
+		Namespace: req.SubjectRelation.Namespace,
+		Relation:  req.SubjectRelation.Relation,
+	}, req.ResourceRelation)
+	if err != nil {
+		return err
+	}
+
+	// For each entrypoint, load the necessary data and re-dispatch if a subproblem was found.
+	return withParallelizedStreamingIterableInCursor(ctx, ci, entrypoints, parentStream, crr.concurrencyLimit,
```

Comment:
We could add it as a tracking field to the dispatch (since we're changing it here anyway) and decrease the counter on each dispatch

Reply:
let's do that as a follow-up, we may want to generalize it for all APIs, and this could be the foundation to what the paper describes as `These safeguards include fine-grained cost accounting, quotas, and throttling`. We could e.g. limit clients by the number of dispatches.